core/src/main/scala/org/apache/spark/storage/BlockManagerDecommissioner.scala

```diff
@@ -66,86 +66,102 @@ private[storage] class BlockManagerDecommissioner(
                 } else {
-                  logInfo(s"Got migration sub-blocks ${blocks}")
-                  logInfo(s"Trying to migrate shuffle ${shuffleBlockInfo} to ${peer} " +
-                    s"($retryCount / $maxReplicationFailuresForDecommission)")
-
-                  // Migrate the components of the blocks.
-                  try {
-                    blocks.foreach { case (blockId, buffer) =>
-                      logDebug(s"Migrating sub-block ${blockId}")
-                      bm.blockTransferService.uploadBlockSync(
-                        peer.host,
-                        peer.port,
-                        peer.executorId,
-                        blockId,
-                        buffer,
-                        StorageLevel.DISK_ONLY,
-                        null) // class tag, we don't need for shuffle
-                      logDebug(s"Migrated sub block ${blockId}")
-                    }
-                    logInfo(s"Migrated ${shuffleBlockInfo} to ${peer}")
-                  } catch {
-                    case e: IOException =>
-                      // If a block got deleted before netty opened the file handle, then trying to
-                      // load the blocks now will fail. This is most likely to occur if we start
-                      // migrating blocks and then the shuffle TTL cleaner kicks in. However this
-                      // could also happen with manually managed shuffles or a GC event on the
-                      // driver a no longer referenced RDD with shuffle files.
-                      if (bm.migratableResolver.getMigrationBlocks(shuffleBlockInfo).isEmpty) {
-                        logWarning(s"Skipping block ${shuffleBlockInfo}, block deleted.")
-                      } else if (fallbackStorage.isDefined) {
-                        fallbackStorage.foreach(_.copy(shuffleBlockInfo, bm))
-                      } else {
-                        throw e
-                      }
-                  }
+                  logError(s"Error occurred during migrating $shuffleBlockInfo", e)
+                  keepRunning = false
                 }
-              } else {
-                logError(s"Skipping block ${shuffleBlockInfo} because it has failed ${retryCount}")
-              }
+              case e: Exception =>
+                logError(s"Error occurred during migrating $shuffleBlockInfo", e)
+                keepRunning = false
+            }
+          }
+          if (keepRunning) {
+            numMigratedShuffles.incrementAndGet()
+          } else {
+            logWarning(s"Stop migrating shuffle blocks to $peer")
+            // Do not mark the block as migrated if it still needs retry
+            if (!allowRetry(shuffleBlockInfo, retryCount + 1)) {
               numMigratedShuffles.incrementAndGet()
+            }
           }
+        } catch {
+          case _: InterruptedException if !keepRunning =>
```

Comment:
In case of `InterruptedException` we should not check the `keepRunning` flag but to stop the thread as soon as possible (especially as it is fatal exception so the next case `NonFatal(e)` won't match for it). Of course we can log a line that we are stopped without requested by the `keepRunning` flag.

Reply:
Good catch! I'll check the `keepRunning` inside the `InterruptedException` and leave an explicit log if the interruption is unexpected.